text/0000-default-field-values.md

```diff
@@ -0,0 +1,1658 @@
+        parse_quote!(#[deny(other_lint)]),
+    ],
+    expr: parse_quote!((1 + 2)),
+    ty: parse_quote!(u8),
+}
+```
+
+## Default fields values are [`const` context]s
+
+[`const` context]: https://github.com/rust-lang-nursery/reference/blob/66ef5396eccca909536b91cad853f727789c8ebe/src/const_eval.md#const-context
+
+As you saw in (7), `Vec::new()`, a function call, was used.
+However, this assumes that `Vec::new` is a *`const fn`*. That is, when you
+provide a default value `field: Type = value`, the given `value` must be a
+*constant expression* such that it is valid in a [`const` context].
+Therefore, you cannot write something like (10):
+
+```rust
+fn launch_missilies() -> Result<(), LaunchFailure> {
+    authenticate()?;
+    begin_launch_sequence()?;
+    ignite()?;
+    Ok(())
+}
+
+struct BadFoo {
+    bad_field: u8 = {
+        launch_missilies().unwrap();
+        42
+    },
+}
+```
+
+Since launching missiles interacts with the real world and has *side-effects*
+in it, it is not possible to do that in a `const` context since it may violate
+deterministic compilation.
+
+## Privacy interactions
+
+In examples thus far, initialization literal expressions `Foo { bar, .. }` have
+been used in the same module as where the type constructed was defined in.
+
+Let's now consider a scenario where this is not the case (11):
+
+```rust
+pub mod foo {
+    pub struct Alpha {
+        beta: u8 = 42,
+        gamma: bool = true,
+    }
+}
+
+mod bar {
+    fn baz() {
+        let x = Alpha { .. };
+    }
+}
+```
+
+Despite `foo::bar` being in a different module than `foo::Alpha` and despite
+`beta` and `gamma` being private to `foo::bar`, a Rust compiler will accept
+the above snippet. It is legal because when `Alpha { .. }` expands to
+`Alpha { beta: 42, gamma: true }`, the fields `beta` and `gamma` are considered
+in the context of `foo::Alpha`'s *definition site* rather than `bar::baz`'s
+definition site.
+
+## `#[non_exhaustive]` interactions
+
+[RFC 2008]: https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#structs-1
+
+[RFC 2008] introduced the attribute `#[non_exhaustive]` that can be placed
+on `struct`, `enum`, and `enum` variants. The RFC notes that upon defining
+a `struct` in *crate A* such as (12):
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub width: u16,
+    pub height: u16,
+}
+```
+
+it is **_not_** possible to initialize a `Config` in a different *crate B* (13):
+
+```rust
+let config = Config { width: 640, height: 480 };
+```
+
+This is forbidden when `#[non_exhaustive]` is attached because the purpose of
+the attribute is to permit adding fields to `Config` without causing a
+breaking change. However, the RFC goes on to note that you can pattern match
+if you allow for the possibility of having fields be ignored with `..` (14):
+
+```rust
+let Config { width, height, .. } = config;
+```
+
+Because this RFC gives a way to have default field values, you can now simply
```

Comment:
This adds the requirement that, in order for a field to be added backwards-compatibly, it has to have a default. I'm not sure whether this is right or not. I think that what we'd want instead is to keep the current banning of iniitlaization syntax for `non_exhaustive`, and users who want the behavior you describe here can add a single private, defaulted field.

Reply:
Added an unresolved question.